sdk/go/common/resource/plugin/provider_test.go

```diff
@@ -131,3 +132,97 @@ func TestNewDetailedDiff(t *testing.T) {
 
 // Assert that UnimplementedProvider implements Provider
 var _ = Provider((*UnimplementedProvider)(nil))
+
+// Regression test for https://github.com/pulumi/pulumi/issues/14335.
+// Ensure that NewDetailedDiffFromObjectDiff builds correct keys.
+func TestNewDetailedDiffFromObjectDiff(t *testing.T) {
+	t.Parallel()
+
+	cases := map[string]struct {
+		diff          *resource.ObjectDiff
+		inputDiff     bool
+		expected      map[string]PropertyDiff
+		expectedPaths map[string]resource.PropertyPath
+	}{
+		"simple add": {
+			diff: &resource.ObjectDiff{
+				Adds: resource.PropertyMap{
+					"a": resource.NewPropertyValue(1),
+				},
+			},
+			expected: map[string]PropertyDiff{
+				"a": {Kind: DiffAdd},
+			},
+			expectedPaths: map[string]resource.PropertyPath{
+				"a": {"a"},
+			},
+		},
+		"simple update": {
+			diff: &resource.ObjectDiff{
+				Updates: map[resource.PropertyKey]resource.ValueDiff{
+					"a": *resource.NewPropertyValue(1).Diff(resource.NewPropertyValue(2)),
+				},
+			},
+			expected: map[string]PropertyDiff{
+				"a": {Kind: DiffUpdate},
+			},
+			expectedPaths: map[string]resource.PropertyPath{
+				"a": {"a"},
+			},
+		},
+		"nested update": {
+			diff: &resource.ObjectDiff{
+				Updates: map[resource.PropertyKey]resource.ValueDiff{
+					"a": *resource.NewObjectProperty(resource.PropertyMap{
+						"b": resource.NewPropertyValue(1),
+					}).Diff(resource.NewObjectProperty(resource.PropertyMap{
+						"b": resource.NewPropertyValue(2),
+					})),
+				},
+			},
+			expected: map[string]PropertyDiff{
+				"a.b": {Kind: DiffUpdate},
+			},
+			expectedPaths: map[string]resource.PropertyPath{
+				"a.b": {"a", "b"},
+			},
+		},
+		"nested update with quoted keys": {
+			diff: &resource.ObjectDiff{
+				Updates: map[resource.PropertyKey]resource.ValueDiff{
+					"a": *resource.NewObjectProperty(resource.PropertyMap{
+						"b.c":            resource.NewPropertyValue(1),
+						"\"quoted key\"": resource.NewPropertyValue(2),
+					}).Diff(resource.NewObjectProperty(resource.PropertyMap{
+						"b.c":            resource.NewPropertyValue(2),
+						"\"quoted key\"": resource.NewPropertyValue(3),
+					})),
+				},
+			},
+			expected: map[string]PropertyDiff{
+				"a[\"\\\"quoted key\\\"\"]": {Kind: DiffUpdate},
+				"a[\"b.c\"]":                {Kind: DiffUpdate},
```

Comment:
Would 
```suggestion
				`a["\"quoted key\""]`: {Kind: DiffUpdate},
				`a["b.c"]`:                {Kind: DiffUpdate},
```

and similar below work here?  That would be a little bit easier to read I think.

Reply:
Yup, can do that.